### PR TITLE
Enable defining custom __new__

### PR DESCRIPTION
**Pull request by Skylion007**
## Description

* This solves the bugs preventing us from defining our own __new__ methods on classes that was brought up in #3253 .

## Suggested changelog entry:

```rst
* Enable defining custom __new__ methods on classes.
* Fixes bug preventing overriding methods if they have non-pybind11 siblings.
```

### CODE REVIEW

**Comment by rwgk** [comment]
How did I arrive at https://github.com/pybind/pybind11/pull/3265/commits/194033e40433fcdfd381837bb228a73629f6584e?

When testing in the Google environment, I ran into this `assert`:

https://github.com/python/cpython/blob/d5650a1738fe34f6e1db4af5f4c4edb7cae90a36/Objects/typeobject.c#L3117

The error wasn't obvious at all, therefore I inserted this right before that `assert`:
```
    if (PyErr_Occurred()) {                                                     
      PyErr_Print();                                                            
      assert(0);                                                                
    } 
```
That way I got to see the `FutureWarning`. After disabling it with the unconditional `#define`, all tests ran ASAN clean.

While looking around, I saw that all new-style `"__init__"` return a pointer (from a placement `new`). Therefore I changed the new code accordingly, before I discovered the `FutureWarning`. I didn't go back to test if it also works when returning a reference (I believe returning a pointer is preferable regardless, although I don't have deep insights).

That's all I know. 🙂

**Comment by Skylion007** [comment]
Defining a custom __new__ function for enums could help close #1177. Also we are putting a class method in the tests now so we might want to add a class method shim as mentioned in #1693 now to py_class as the current way of defining class methods is bulky and not very intuitive. 

We don't seem to have a way to check the py::type of a the C++ class for the first argument though, which is a bit annoying.

**Comment by Skylion007** [comment]
I think even if we reduce the scope to calling attr("get") it would be a useful shim.

**Comment by Skylion007** [comment]
This syntax is at the very least not terrible until we have proper classmethod support.

File: include/pybind11/pybind11.h [hunk 1]
```diff
@@ -416,7 +416,13 @@ class cpp_function : public function {
...
         if (rec->sibling) {
             if (PyCFunction_Check(rec->sibling.ptr())) {
-                auto rec_capsule = reinterpret_borrow<capsule>(PyCFunction_GET_SELF(rec->sibling.ptr()));
+                auto *self = PyCFunction_GET_SELF(rec->sibling.ptr());
+                capsule rec_capsule;
+                if (isinstance<capsule>(self)) {
+                    rec_capsule = reinterpret_borrow<capsule>(self);
+                } else {
+                    rec_capsule = capsule(self);
+                }
```
Comment (henryiii) [inline include/pybind11/pybind11.h:425]:
```suggestion
                rec_capsule = isinstance<capsule>(self) ? reinterpret_borrow<capsule>(self) : capsule(self);
```

Comment (henryiii) [inline include/pybind11/pybind11.h:425]:
```suggestion
                capsule rec_capsule = isinstance<capsule>(self) ? reinterpret_borrow<capsule>(self) : capsule(self);
```

**Review by henryiii — APPROVED** [review]

**Comment by Skylion007** [comment]
I actually just realized those PTRs in test should probably be ref since they can't be null. Probably can go back and fix them when we write the actual documentation for this feature at some point @henryiii along with proper classmethod support perhaps

File: tests/test_class.cpp [hunk 2]
```diff
@@ -47,10 +47,25 @@ TEST_SUBMODULE(class_, m) {
...
+            return ptr;
+        }
+        ~NoConstructorNew() { print_destroyed(this); }
+    };
 
     py::class_<NoConstructor>(m, "NoConstructor")
         .def_static("new_instance", &NoConstructor::new_instance, "Return an instance");
 
+    py::class_<NoConstructorNew>(m, "NoConstructorNew")
+        .def(py::init([](NoConstructorNew *self) { return self; })) // Need a NOOP __init__
```
Comment (Skylion007) [inline tests/test_class.cpp:66]: I also wish there was a cleaner way to not have to implement `__init__` if __new__ is defined, but I can't figure out how to hack to the metaclass not to complain.

Comment (rwgk) [reply tests/test_class.cpp:66]: Maybe something like `.def_new(...)` or `.def(py::new(...))` that wraps the two methods you have right now?

**Pull request merged**